ParseCAN/spec/bus/frame.py

```diff
@@ -165,5 +189,36 @@ def __post_init__(self):
 
         self.frame.extend(frame)
 
-    def pack(self):
-        raise NotImplementedError()
+    def unpack(self, frame, **kwargs):
+        mux_id = frame[self.slice.start, self.slice.length]
+        return (self.name,
+                self.frame['key'][mux_id].unpack(frame, **kwargs))
+
+    def pack(self, id_tup, bitstring=0, length=None, by='name', **kwargs):
+        # get top level length
+        if length is None:
+            length = len(self)
+
+        name = id_tup[0]
+        id_tup = id_tup[1]
+        bitstring = data.evil_macros.INSERT(
+                self.frame[by][name].key,
+                bitstring,
+                self.slice.start,
+                self.slice.length
+            )
+        frame = self.frame[by][name]
+        return frame.pack(id_tup=id_tup,
+                            bitstring=bitstring,
+                            length=length,
+                            by=by,
+                            **kwargs)
+
+    def __len__(self):
+        # paranoid length
+        # return ceil(
+        #             max(self.slice.start+self.slice.length, 
+        #                   max(len(frame) for frame in self.frame))/8
+        #         )
+        # assuming mux id is at end of frame
```

Comment:
It doesn't have to be. In fact it is usually at the start.

Reply:
We can memoize this for now as we are adding the sub frames, even though it's unsafe considering immutability.